Request Deployment with array as input

**vRA version**
8.1
**Terraform version**
Terraform v0.12.25
**terraform-provider-vra plugin version**
provider.vra v0.2.0

**Describe the bug**
When trying to apply terraform config for deploying a blueprint which has an input of type array of strings terraform fails. 
based on the go code for input schema on resource_deployment, it looks like it expects only strings.

**To Reproduce**
Steps to reproduce the behavior:
1- create a blueprint that receives an input of array of strings 
2- create terraform files to request the deployment from the blueprint  
3- terraform plan fails with the error "string required"

**Expected behavior**
run smoothly :D

Hello,

Just a comment to say that I also have the same problem with an input for the disk size that I want to update from terraform.

```
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: -----------------------------------------------------
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 HTTP/1.1 400 Bad Request
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Content-Length: 112
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Cache-Control: no-cache, no-store, max-age=0, must-revalidate
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Content-Type: application/json
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Date: Fri, 10 Jul 2020 12:29:57 GMT
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Expires: 0
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Pragma: no-cache
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 Strict-Transport-Security: max-age=31536000 ; includeSubDomains
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 X-Content-Type-Options: nosniff
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 X-Frame-Options: DENY
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 X-Xss-Protection: 1; mode=block
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58 {"message":"Invalid inputs: diskSize: Expected type NUMBER but was STRING..","statusCode":400,"errorCode":20021}
2020-07-10T14:29:58.452+0200 [DEBUG] plugin.terraform-provider-vra_v0.2.2.exe: 2020/07/10 14:29:58
2020/07/10 14:29:58 [DEBUG] vra_deployment.this: apply errored, but we're indicating that via the Error pointer rather than returning it: unknown error (status 400): {resp:0xc000385320}
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalWriteState
2020/07/10 14:29:58 [TRACE] EvalWriteState: recording 0 dependencies for vra_deployment.this
2020/07/10 14:29:58 [TRACE] EvalWriteState: writing current state object for vra_deployment.this
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalApplyProvisioners
2020/07/10 14:29:58 [TRACE] EvalApplyProvisioners: vra_deployment.this is not freshly-created, so no provisioning is required
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalMaybeTainted
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalWriteState
2020/07/10 14:29:58 [TRACE] EvalWriteState: recording 0 dependencies for vra_deployment.this
2020/07/10 14:29:58 [TRACE] EvalWriteState: writing current state object for vra_deployment.this
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalIf
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalIf
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalWriteDiff
2020/07/10 14:29:58 [TRACE] <root>: eval: *terraform.EvalApplyPost
2020/07/10 14:29:58 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: unknown error (status 400): {resp:0xc000385320}
2020/07/10 14:29:58 [ERROR] <root>: eval: *terraform.EvalSequence, err: unknown error (status 400): {resp:0xc000385320}
2020/07/10 14:29:58 [TRACE] [walkApply] Exiting eval tree: vra_deployment.this
2020/07/10 14:29:58 [TRACE] vertex "vra_deployment.this": visit complete
2020/07/10 14:29:58 [TRACE] dag/walk: upstream of "provider.vra (close)" errored, so skipping
2020/07/10 14:29:58 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2020/07/10 14:29:58 [TRACE] dag/walk: upstream of "root" errored, so skipping
2020/07/10 14:29:58 [TRACE] statemgr.Filesystem: not making a backup, because the new snapshot is identical to the old
2020/07/10 14:29:58 [TRACE] statemgr.Filesystem: no state changes since last snapshot
2020/07/10 14:29:58 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2020/07/10 14:29:58 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info

Error: unknown error (status 400): {resp:0xc000385320}

  on main.tf line 16, in resource "vra_deployment" "this":
  16: resource "vra_deployment" "this" {

2020/07/10 14:29:58 [TRACE] statemgr.Filesystem: unlocked by closing terraform.tfstate
```

On my terraform code, the input "diskSize" is a number but it seems that terraform always passes inputs as string.

--- a/examples/deployment/blueprint/main.tf
+++ b/examples/deployment/blueprint/main.tf
@@ -22,8 +22,10 @@ resource "vra_deployment" "this" {
   inputs = {
     flavor = "small"
     image  = "centos"
-    count  = 2
+    count  = 1
     flag   = true
+    arrayProp = jsonencode(["foo", "bar", "baz"])
+    objectProp = jsonencode({ "key": "value", "key2": [1, 2, 3] })
   }
 
   timeouts {

--- a/examples/deployment/catalog_item/main.tf
+++ b/examples/deployment/catalog_item/main.tf
@@ -26,6 +26,9 @@ resource "vra_deployment" "this" {
     image  = "centos"
     count  = 1
     flag   = false
+    number = 10.0
+    arrayProp = jsonencode(["foo", "bar", "where", "waldo"])
+    objectProp = jsonencode({ "key1": "value1", "key2": [1, 2, 3, 4] })
   }
 
   timeouts {

--- a/vra/deployment_request.go
+++ b/vra/deployment_request.go
@@ -151,7 +151,7 @@ func flattenDeploymentRequest(deploymentRequest *models.DeploymentRequest) inter
 	helper["dismissed"] = deploymentRequest.Dismissed
 	helper["id"] = deploymentRequest.ID.String()
 	helper["initialized_at"] = deploymentRequest.InitializedAt.String()
-	helper["inputs"] = expandInputs(deploymentRequest.Inputs)
+	helper["inputs"] = expandInputsToString(deploymentRequest.Inputs)
 	helper["name"] = deploymentRequest.Name
 	helper["parent_id"] = deploymentRequest.ParentID.String()
 	helper["requested_by"] = deploymentRequest.RequestedBy

--- a/vra/resource_deployment.go
+++ b/vra/resource_deployment.go
@@ -1,6 +1,7 @@
 package vra
 
 import (
+	"encoding/json"
 	"fmt"
 	"strconv"
 
@@ -242,6 +243,8 @@ func resourceDeploymentCreate(d *schema.ResourceData, m interface{}) error {
 		}
 
 		if v, ok := d.GetOk("inputs"); ok {
+			// If the inputs are provided, get the schema from blueprint to convert the provided input values
+			// to the type defined in the schema.
 			inputs, err = getBlueprintInputsByType(apiClient, blueprintID, blueprintVersion, v)
 			if err != nil {
 				return err
@@ -310,6 +313,9 @@ func resourceDeploymentRead(d *schema.ResourceData, m interface{}) error {
 	log.Printf("Reading the vra_deployment resource with name %s", d.Get("name"))
 	apiClient := m.(*Client).apiClient
 
+	// Getting the input types map
+	inputTypesMap := getInputTypesMap(d, apiClient)
+
 	id := d.Id()
 	expandProject := d.Get("expand_project").(bool)
 
@@ -343,14 +349,14 @@ func resourceDeploymentRead(d *schema.ResourceData, m interface{}) error {
 		return fmt.Errorf("error setting deployment expense - error: %#v", err)
 	}
 
-	allInputs := expandInputs(deployment.Inputs)
-	if err := d.Set("inputs_including_defaults", allInputs); err != nil {
+	if err := d.Set("inputs_including_defaults", expandInputsToString(deployment.Inputs)); err != nil {
 		return fmt.Errorf("error setting deployment inputs_including_defaults - error: %#v", err)
 	}
 
+	allInputs := expandInputs(deployment.Inputs)
 	if v, ok := d.GetOk("inputs"); ok {
 		userInputs := v.(map[string]interface{})
-		if err := d.Set("inputs", updateUserInputs(allInputs, userInputs)); err != nil {
+		if err := d.Set("inputs", updateUserInputs(allInputs, userInputs, inputTypesMap)); err != nil {
 			return fmt.Errorf("error setting deployment inputs - error: %#v", err)
 		}
 	}
@@ -378,6 +384,46 @@ func resourceDeploymentRead(d *schema.ResourceData, m interface{}) error {
 
 	log.Printf("Finished reading the vra_deployment resource with name '%s'. Current status: '%s'", d.Get("name"), d.Get("status"))
 	return nil
+}
+
+// Gets the inputs and their types as map[string]string
+func getInputTypesMap(d *schema.ResourceData, apiClient *client.MulticloudIaaS) map[string]string {
+	inputTypesMap := make(map[string]string)
+
+	if _, ok := d.GetOk("inputs"); !ok {
+		return inputTypesMap
+	}
+
+	// Get blueprint_id and catalog_item_id
+	blueprintID, catalogItemID := "", ""
+	if v, ok := d.GetOk("blueprint_id"); ok {
+		blueprintID = v.(string)
+	}
+
+	if v, ok := d.GetOk("catalog_item_id"); ok {
+		catalogItemID = v.(string)
+	}
+
+	if catalogItemID != "" {
+		// Get the catalog item inputs and their types
+		catalogItemVersion := ""
+		if v, ok := d.GetOk("catalog_item_version"); ok {
+			catalogItemVersion = v.(string)
+		}
+
+		inputTypesMap, _ = getCatalogItemInputTypesMap(apiClient, catalogItemID, catalogItemVersion)
+	} else if blueprintID != "" {
+		// Get the blueprint inputs and their types
+		blueprintVersion := ""
+		if v, ok := d.GetOk("blueprint_version"); ok {
+			blueprintVersion = v.(string)
+		}
+
+		inputTypesMap, _ = getBlueprintInputTypesMap(apiClient, blueprintID, blueprintVersion)
+	}
+
+	log.Printf("InputTypesMap: %v", inputTypesMap)
+	return inputTypesMap
 }
 
 func resourceDeploymentUpdate(d *schema.ResourceData, m interface{}) error {
@@ -461,13 +507,7 @@ func resourceDeploymentDelete(d *schema.ResourceData, m interface{}) error {
 }
 
 func getCatalogItemInputsByType(apiClient *client.MulticloudIaaS, catalogItemID string, catalogItemVersion string, inputValues interface{}) (map[string]interface{}, error) {
-	inputsSchemaMap, err := getCatalogItemSchema(apiClient, catalogItemID, catalogItemVersion)
-	if err != nil {
-		return nil, err
-	}
-
-	log.Printf("Inputs Schema: %v", inputsSchemaMap)
-	inputTypesMap, err := getInputTypesMapFromCatalogItemSchema(inputsSchemaMap)
+	inputTypesMap, err := getCatalogItemInputTypesMap(apiClient, catalogItemID, catalogItemVersion)
 	if err != nil {
 		return nil, err
 	}
@@ -481,14 +521,23 @@ func getCatalogItemInputsByType(apiClient *client.MulticloudIaaS, catalogItemID 
 	return inputs, nil
 }
 
-func getBlueprintInputsByType(apiClient *client.MulticloudIaaS, blueprintID string, blueprintVersion string, inputValues interface{}) (map[string]interface{}, error) {
-	inputsSchemaMap, err := getBlueprintSchema(apiClient, blueprintID, blueprintVersion)
+func getCatalogItemInputTypesMap(apiClient *client.MulticloudIaaS, catalogItemID string, catalogItemVersion string) (map[string]string, error) {
+	log.Printf("Getting Catalog Item Schema for catalog_item_id: [%v], catalog_item_version: [%v]", catalogItemID, catalogItemVersion)
+	inputsSchemaMap, err := getCatalogItemSchema(apiClient, catalogItemID, catalogItemVersion)
 	if err != nil {
 		return nil, err
 	}
 
 	log.Printf("Inputs Schema: %v", inputsSchemaMap)
-	inputTypesMap, err := getInputTypesMapFromBlueprintInputsSchema(inputsSchemaMap)
+	inputTypesMap, err := getInputTypesMapFromCatalogItemSchema(inputsSchemaMap)
+	if err != nil {
+		return nil, err
+	}
+	return inputTypesMap, nil
+}
+
+func getBlueprintInputsByType(apiClient *client.MulticloudIaaS, blueprintID string, blueprintVersion string, inputValues interface{}) (map[string]interface{}, error) {
+	inputTypesMap, err := getBlueprintInputTypesMap(apiClient, blueprintID, blueprintVersion)
 	if err != nil {
 		return nil, err
 	}
@@ -500,6 +549,21 @@ func getBlueprintInputsByType(apiClient *client.MulticloudIaaS, blueprintID stri
 	}
 
 	return inputs, nil
+}
+
+func getBlueprintInputTypesMap(apiClient *client.MulticloudIaaS, blueprintID string, blueprintVersion string) (map[string]string, error) {
+	log.Printf("Getting Blueprint Schema for blueprint_id: [%v], blueprint_version: [%v]", blueprintID, blueprintVersion)
+	inputsSchemaMap, err := getBlueprintSchema(apiClient, blueprintID, blueprintVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("Inputs Schema: %v", inputsSchemaMap)
+	inputTypesMap, err := getInputTypesMapFromBlueprintInputsSchema(inputsSchemaMap)
+	if err != nil {
+		return nil, err
+	}
+	return inputTypesMap, nil
 }
 
 func getCatalogItemSchema(apiClient *client.MulticloudIaaS, catalogItemID string, catalogItemVersion string) (map[string]interface{}, error) {
@@ -791,12 +855,28 @@ func getInputsByType(inputs map[string]interface{}, inputTypesMap map[string]str
 	var err error
 	for k, v := range inputs {
 		if t, ok := inputTypesMap[k]; ok {
-			log.Printf("input_key: %s, type: %#v", k, t)
+			log.Printf("input_key: %s, type: %#v, value provided: %#v", k, t, v)
 			switch strings.ToLower(t) {
+			case "array":
+				value := make([]interface{}, 0)
+				err := json.Unmarshal([]byte(v.(string)), &value)
+				if err != nil {
+					return nil, fmt.Errorf("cannot convert input '%v' value '%v' into type '%s'. %#v", k, v, t, err)
+				}
+				inputsByType[k] = value
 			case "boolean":
 				inputsByType[k], err = strconv.ParseBool(v.(string))
 			case "integer":
 				inputsByType[k], err = strconv.Atoi(v.(string))
+			case "number":
+				inputsByType[k], err = strconv.ParseFloat(v.(string), 32)
+			case "object":
+				var value interface{}
+				err := json.Unmarshal([]byte(v.(string)), &value)
+				if err != nil {
+					return nil, fmt.Errorf("cannot convert input '%v' value '%v' into type '%s'. %#v", k, v, t, err)
+				}
+				inputsByType[k] = value
 			default:
 				inputsByType[k] = v
 			}
@@ -838,7 +918,9 @@ func runAction(d *schema.ResourceData, apiClient *client.MulticloudIaaS, deploym
 	}
 
 	resp, err := apiClient.DeploymentActions.SubmitDeploymentActionRequestUsingPOST(
-		deployment_actions.NewSubmitDeploymentActionRequestUsingPOSTParams().WithDepID(deploymentUUID).
+		deployment_actions.NewSubmitDeploymentActionRequestUsingPOSTParams().
+			WithAPIVersion(withString(DeploymentsAPIVersion)).
+			WithDepID(deploymentUUID).
 			WithActionRequest(&resourceActionRequest))
 	if err != nil {
 		return err
@@ -905,18 +987,40 @@ func deploymentDeleteStatusRefreshFunc(apiClient client.MulticloudIaaS, id strin
 	}
 }
 
-// Updates only user inputs from all the inputs
-func updateUserInputs(allInputs, userInputs map[string]interface{}) map[string]interface{} {
+func updateUserInputs(allInputs, userInputs map[string]interface{}, inputTypesMap map[string]string) map[string]interface{} {
 	if allInputs == nil || userInputs == nil {
 		return nil
 	}
 
 	inputs := make(map[string]interface{})
-	for key, value := range userInputs {
+	for name, value := range userInputs {
 		if value != nil {
-			inputs[key] = fmt.Sprint(allInputs[key])
+			inputs[name] = decodeInputValue(name, allInputs[name], inputTypesMap)
 		}
+		log.Printf("Converted incoming value to string: Key: %v, Value: %v, Converted value: %#v", name, allInputs[name], inputs[name])
 	}
 
 	return inputs
+}
+
+func decodeInputValue(inputName string, inputValue interface{}, inputTypesMap map[string]string) interface{} {
+	if t, ok := inputTypesMap[inputName]; ok {
+		log.Printf("input_key: %s, type: %#v, value: %#v", inputName, t, inputValue)
+		switch strings.ToLower(t) {
+		case "array":
+			fallthrough
+		case "object":
+			log.Printf("Converting input '%v' of type '%v'", inputName, t)
+			value, err := json.Marshal(inputValue)
+			if err != nil {
+				log.Printf("cannot convert input '%v' value '%v' into type '%v'. %#v", inputName, inputValue, t, err)
+				return fmt.Sprint(inputValue)
+			}
+			return string(value)
+		default:
+			return fmt.Sprint(inputValue)
+		}
+	}
+
+	return fmt.Sprint(inputValue)
 }

--- a/vra/structure.go
+++ b/vra/structure.go
@@ -54,6 +54,24 @@ func compareUnique(s []interface{}) bool {
 	return j == len(s)
 }
 
+////diff returns the elements in 's1' that are not in 's2'
+//func diff(s1, s2 []interface{}) []interface{} {
+//	m2 := make(map[string]struct{}, len(s2))
+//
+//	for _, v := range s2 {
+//		m2[v.(string)] = struct{}{}
+//	}
+//
+//	var diff []interface{}
+//	for _, v := range s1 {
+//		if _, found := m2[v.(string)]; !found {
+//			diff = append(diff, v)
+//		}
+//	}
+//
+//	return diff
+//}
+
 // indexOf will lookup and return the index of value in the list of items
 func indexOf(value string, items []string) (int, error) {
 	for i, v := range items {
@@ -149,13 +167,30 @@ func flattenAndNormalizeCLoudAccountGcpRegionIds(regionOrder []string, cloudAcco
 	return m, nil
 }
 
-// expandInputs will convert the interface  into a map of interface
+// expandInputs will convert the interface  into a map of [string:interface]
 func expandInputs(configInputs interface{}) map[string]interface{} {
 	if configInputs == nil {
 		return nil
 	}
 
 	inputs := make(map[string]interface{})
+	for key, value := range configInputs.(map[string]interface{}) {
+		if value != nil {
+			//inputs[key] = fmt.Sprint(value)
+			inputs[key] = value
+		}
+	}
+
+	return inputs
+}
+
+// expandInputsToString will convert the interface  into a map of string:string
+func expandInputsToString(configInputs interface{}) map[string]string {
+	if configInputs == nil {
+		return nil
+	}
+
+	inputs := make(map[string]string)
 	for key, value := range configInputs.(map[string]interface{}) {
 		if value != nil {
 			inputs[key] = fmt.Sprint(value)

--- a/vra/structure.go
+++ b/vra/structure.go
@@ -54,24 +54,6 @@ func compareUnique(s []interface{}) bool {
 	return j == len(s)
 }
 
-////diff returns the elements in 's1' that are not in 's2'
-//func diff(s1, s2 []interface{}) []interface{} {
-//	m2 := make(map[string]struct{}, len(s2))
-//
-//	for _, v := range s2 {
-//		m2[v.(string)] = struct{}{}
-//	}
-//
-//	var diff []interface{}
-//	for _, v := range s1 {
-//		if _, found := m2[v.(string)]; !found {
-//			diff = append(diff, v)
-//		}
-//	}
-//
-//	return diff
-//}
-
 // indexOf will lookup and return the index of value in the list of items
 func indexOf(value string, items []string) (int, error) {
 	for i, v := range items {

--- a/vra/structure_test.go
+++ b/vra/structure_test.go
@@ -1,7 +1,6 @@
 package vra
 
 import (
-	"fmt"
 	"testing"
 )
 
@@ -22,7 +21,7 @@ func TestExpandInputs(t *testing.T) {
 
 	expandedInputs := expandInputs(inputs)
 
-	if expandedInputs["whole"] != fmt.Sprint(2) {
+	if expandedInputs["whole"] != 2 {
 		t.Errorf("int type input is not expanded correctly.")
 	}
 
@@ -30,15 +29,15 @@ func TestExpandInputs(t *testing.T) {
 		t.Errorf("string type input is not expanded correctly.")
 	}
 
-	if expandedInputs["flag"] != fmt.Sprint(true) {
+	if expandedInputs["flag"] != true {
 		t.Errorf("bool type input is not expanded correctly.")
 	}
 
-	if expandedInputs["fraction"] != "2.4" {
+	if expandedInputs["fraction"] != 2.4 {
 		t.Errorf("float type input is not expanded correctly.")
 	}
 
-	if expandedInputs["message"] != fmt.Sprint(m) {
-		t.Errorf("float type input is not expanded correctly.")
+	if expandedInputs["message"] != m {
+		t.Errorf("object type input is not expanded correctly.")
 	}
 }

